main/src/library/RedBlackTree.flix

```diff
@@ -396,4 +398,31 @@ namespace RedBlackTree {
         case _ => tree
     }
 
+    ///
+    /// Extracts a range of key-value pairs from `tree`.
+    ///
+    /// That is, the result is a list of all pairs `f(k, v)` where `p(k)` returns `Equal`.
+    ///
+    pub def queryWith(p: k -> Ordering & e, f: (k, v) -> a, tree: RedBlackTree[k, v]): List[a] & e =
+        let buffer = MutList.new() as & e;
+        queryWithHelper(p, f, buffer, tree) as & e;
+        MutList.foldRight((x, acc) -> x :: acc, Nil, buffer) as & e
```

Comment:
I think we should have: `MutList.toImmutable()`. Feel free to add that in this PR.